common/network-common/src/main/java/org/apache/spark/network/util/TransportConf.java

```diff
@@ -111,8 +111,30 @@ public int numConnectionsPerPeer() {
   /** Requested maximum length of the queue of incoming connections. Default is 64. */
   public int backLog() { return conf.getInt(SPARK_NETWORK_IO_BACKLOG_KEY, 64); }
 
-  /** Number of threads used in the server thread pool. Default to 0, which is 2x#cores. */
-  public int serverThreads() { return conf.getInt(SPARK_NETWORK_IO_SERVERTHREADS_KEY, 0); }
+  /**
+   * The configured ratio between number of server threads and number of chunk fetch handler
+   * threads. Default to 1, which sets the size of both thread pools to be equal. Number of
+   * server threads needs to be a multiple of this ratio. See SPARK-29206.
+   */
+  private int getChunkFetchHandlerThreadsRatio() {
+    return conf.getInt("spark.shuffle.server.chunkFetchHandlerThreadsRatio", 1);
+  }
+
+  /**
+   * Number of threads used in the server thread pool. Default to 0, which is 2x#cores.
+   * If spark.shuffle.server.chunkFetchHandlerThreadsRatio is configured, and the Netty server
+   * is for shuffle, then the actual # of server threads will round up to the nearest int that
+   * is a multiple of the configured ratio.
+   */
+  public int serverThreads() {
+    int configuredServerThreads = conf.getInt(SPARK_NETWORK_IO_SERVERTHREADS_KEY, 0);
+    if (this.getModuleName().equalsIgnoreCase("shuffle")) {
+      int chunkFetchHandlerThreadsRatio = getChunkFetchHandlerThreadsRatio();
+      return (int) Math.ceil(configuredServerThreads / (chunkFetchHandlerThreadsRatio * 1.0));
```

Comment:
This logic  seems wrong. like I said above the Ratio is an int.
Is also doesn't work with the default configuredServerThreads being 0.  0/X = 0
If we are leaving the configs like they were before you don't need to change serverThreads. That is what user configured or default and then the ratio/percent fetch handler config applies to that.

Reply:
If we leave the config like before, user could configure serverThreads as 9, and the percent as 50%. It would then use 9 server threads and 5 chunk fetch handler threads. Since 9 is not a multiple of 5, with this config we might experience RPC message timeout issue again.

With the ratio config, user could configure serverThreads as 9 and ratio as 2. Then, with this patch, we would configure 10 server threads and 5 chunk fetch handler threads. This would guarantee to avoid the issue we saw.

It still works with the default case as serverThreads() will still return 0 here. The # chunk fetch handler threads will then be configured as 2 * # cores, same as the number of server threads. This is the same as the current behavior.

Reply:
right the server threads stay 0 but the chunk fetcher threads stays 0 as well, that is different than the previous config, which would still take 2*availablecores * chunkFetchHandlerThreadsPercent/100.
so this requires you to set server threads for the config to work whereas previously it didn't. I would like to keep the behavior of it applying even if don't explicitly set server threads.

the logic here also still seems wrong but maybe I'm still missing what your config is.

I configure server threads to 10 and thread ratio to 2. 
(int) Math.ceil(configuredServerThreads / (chunkFetchHandlerThreadsRatio * 1.0)) = 5 server threads which is not close to what I set the config to (10).  I would have expected server threads to stay 10 and the chunk fetch handler threads to be 5.  You are missing a multiply back by chunkFetchHandlerThreadsRatio.

Reply:
If the server threads is 0, the logic in this patch would configure the number of chunk fetch handler threads to be 2 * # cores. The current default behavior would set server threads to 2 * # cores and set the percent to 100%. So the number of chunk fetch handler threads is also 2 * # cores.

Line 362-363 in this patch:
return this.serverThreads() > 0 ? this.serverThreads() / chunkFetchHandlerThreadsRatio :
    2 * NettyRuntime.availableProcessors();

Reply:
I understand and agree if you don't set anything then the default is the same, but In the existing code before this change, if I don't set server threads and I set spark.shuffle.server.chunkFetchHandlerThreadsPercent=80, then the code below will apply it: (assume available processors =16)

 (logic below in line 337-341):

int chunkFetchHandlerThreadsPercent = 80
int threads = 2 * NettyRuntime.availableProcessors(); = 32
return (int) Math.ceil(threads * (chunkFetchHandlerThreadsPercent / 100.0)); = 26

thus chunk fetcher threads is 26 and server threads is 32. Your patch here doesn't keep similar behavior. The only way for the config spark.shuffle.server.chunkFetchHandlerThreadsRatio to apply is by also setting the server threads explicitly.

Again you also need to fix your rounding code above, you are missing the multiply.

Reply:
The reason for not applying spark.shuffle.server.chunkFetchHandlerThreadsRatio when server threads is not explicitly configured is because the behavior to set # server threads to 2 * cores is inside Netty. We cannot guarantee 2 * cores can be divided by the int_ratio or multiplier for a better name, unless we also overwrite the # server threads when it's not configured.

I was previously thinking maybe the default behavior of server threads being 2 * cores should be preserved, but I think you are right in that if spark.shuffle.server.chunkFetchHandlerThreadsRatio is configured, it should be honored.